R/new.R

```diff
@@ -51,18 +56,33 @@ new_valid_tibble <- function(x, nrow, subclass = NULL) {
   #' @details
   #' 1d arrays are always converted to vectors.
   x[] <- map(x, strip_dim)
+  set_tibble_class(x, ..., .nrow = .nrow, .subclass = .subclass)
+}
+
+tibble_classes <- c("tbl_df", "tbl", "data.frame")
+
+#' @rdname new_tibble
+#' @usage NULL
+set_tibble_class <- function(x, ..., .nrow, .subclass) {
+  attribs <- list2(...)
```

Comment:
I think the code specifically related to tibble attributes should be inlined in `new_tibble()`.

Then you should have another function like this:

```R
vec_restore.data.frame <- function(x, to) {
  # Copy attribute, preserving existing names & recreating rownames
  attr_to <- attributes(to)
  attr_to[["names"]] <- names(x)
  attr_to[["row.names"]] <- .set_row_names(df_length(x))
  attributes(x) <- attr_to

  x
}
```

i.e. it needs to take the other object as an argument, not just a list of attributes.